Add JSON path example

akupila opened the pull request:
As i was [stumbling with the basics](https://github.com/alecthomas/participle/issues/72), hopefully a super simple example can help others get started. 

This PR adds a new example for parsing a json path and applying that to a json file, a bit like an interactive but primitive version of [jq](https://stedolan.github.io/jq/).

Demo: https://asciinema.org/a/nSWGX6ZPghlpG8blXwGvYmui5

alecthomas commented on _examples/json/main.go:
```diff
@@ -0,0 +1,109 @@
+
+	for {
+		// Example: check_run.check_suite.pull_requests[0].head.sha
+		fmt.Fprint(os.Stderr, "Enter a json path: ")
```
I'd just make this an argument ge. `json <query> <files...>`

akupila replied:
I made it so it's not interactive any more. Instead you pass in a query and 1+ files

alecthomas commented on _examples/json/main.go:
```diff
@@ -0,0 +1,109 @@
+
+		var expr pathExpr
+		if err := parser.ParseString(q, &expr); err != nil {
+			fmt.Fprintln(os.Stderr, err)
```
Use `ctx.FatalIfErrorf(err)` here.

akupila replied:
The idea previously was that the app should not exit on error, but instead ask for a new query. Now however as it's no longer interactive, i removed the kong dependency to keep thing simpler and just do `os.Exit(1)` instead.

alecthomas commented on _examples/json/main.go:
```diff
@@ -0,0 +1,109 @@
+		}
+
+		result, err := match(input, expr)
+		if err != nil {
```
Same here.

alecthomas commented on _examples/json/main.go:
```diff
@@ -0,0 +1,109 @@
+		if v.Kind() == reflect.Interface {
+			v = v.Elem()
+		}
+		for _, a := range e.Acc {
```
Nice!

alecthomas commented on _examples/json/main.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+func match(input map[string]interface{}, expr pathExpr) (interface{}, error) {
+	v := reflect.ValueOf(input)
```
I'd probably just use type switches here, but up to you.

akupila replied:
Done. Not sure if i find the non-reflect code easier to read but maybe it doesn't stand out as much to however is reading the example.

akupila commented:
I addressed your comments with a couple changes to how this works:
- It is no longer interactive. Instead you pass in the the query and files to query and it'll execute it (more like `jq`). For example `go run . "check_run.check_suite.pull_requests[0].head.sha" github-webhook.json`
- As it is no longer interactive, any error will immediately terminate
- The matching is done with type switches rather than using the reflect package.
- The output is encoded as json only if the match is for an object. Not important for the example but i find this a bit nicer to use

akupila commented on _examples/json/main.go:
```diff
@@ -0,0 +1,114 @@
+}
+
+type acc struct {
+	Name  *string `@(String|Char|RawString)`
```
Is this correct? Would `@String` be sufficient?

alecthomas replied:
Yep `@String` should be sufficient.

alecthomas commented:
Thanks!

itchyny commented:
How about renaming the directory to `json_path` so that the visitor does not confuse the example parses JSON? And maybe adding an example for JSON parser?
